### PR TITLE
feat: Replace petgraph graph with custom graph library

### LINKED ISSUE #12: Refactor graph structure of medrecord

### PR DESCRIPTION
**Pull request by JabobKrauskopf**
Closes #12 #31 and #23

### CODE REVIEW

**Review by philippkochlimebit — DISMISSED** [review]

File: medmodels/medrecord/tests/test_medrecord.py [hunk 1]
```diff
@@ -182,7 +182,11 @@ def test_node(self):
```
[the remark this answers is not in the document]
Comment (MarIniOnz) [reply medmodels/medrecord/tests/test_medrecord.py:182]: Add in Line 101 a test for the case there are more than 2 indices.

**Review by MarIniOnz — CHANGES_REQUESTED** [review]
Overall, very nice. Requested some minor changes in Python files and tests.

File: medmodels/_medmodels.pyi [hunk 2]
```diff
@@ -27,24 +36,36 @@ class PyMedRecord:
...
+        self, group: Group, node_indices_to_add: Optional[List[NodeIndex]]
     ) -> None: ...
-    def add_group(self, group: str, node_ids_to_add: Optional[List[str]]) -> None: ...
-    def remove_group(self, group: str) -> None: ...
-    def remove_from_group(self, group: str, node_id: str) -> None: ...
-    def add_to_group(self, group: str, node_id: str) -> None: ...
-    def neighbors(self, *node_id: str) -> List[tuple[str, Dict]]: ...
+    def remove_group(self, group: Group) -> None: ...
+    def remove_from_group(self, group: Group, node_index: NodeIndex) -> None: ...
+    def add_to_group(self, group: Group, nodex_index: NodeIndex) -> None: ...
```
[the remark this answers is not in the document]
Comment (JabobKrauskopf) [reply medmodels/_medmodels.pyi:70]: thank you and i agree :)

File: medmodels/_medmodels.pyi [hunk 3]
```diff
@@ -27,24 +36,36 @@ class PyMedRecord:
...
-    def add_node(self, id: str, attributes: Dict) -> None: ...
-    def add_nodes(self, nodes: List[tuple[str, Dict]]) -> None: ...
+    def node(self, *node_index: NodeIndex) -> Dict[NodeIndex, Attributes]: ...
+    def edge(self, *edge_index: EdgeIndex) -> Dict[EdgeIndex, Attributes]: ...
+    def edges_connecting(
+        self, source_node_index: NodeIndex, target_node_index: NodeIndex
+    ) -> List[EdgeIndex]: ...
+    def group(self, *group: Group) -> Dict[Group, List[NodeIndex]]: ...
+    def add_node(self, node_index: NodeIndex, attributes: Attributes) -> None: ...
+    def add_nodes(self, nodes: List[tuple[NodeIndex, Attributes]]) -> None: ...
```
[the remark this answers is not in the document]
Comment (JabobKrauskopf) [reply medmodels/_medmodels.pyi:46]: then each node would have to be wrapped in a tuple `def add_node(self, *node: tuple[NodeIndex, Attributes])`. I'd prefer the current interface where you have both options

**Review by MarIniOnz — APPROVED** [review]
Comments solved or explained. Well done!

**Review by FloLimebit — APPROVED** [review]
awesometastic PR!

**Review by philippkochlimebit — APPROVED** [review]

**Pull request merged**